// @ark-ui/react/dist/components/scroll-area/index.cjs
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const scrollAreaContext = require('./scroll-area-context.cjs');
const scrollAreaContent = require('./scroll-area-content.cjs');
const scrollAreaCorner = require('./scroll-area-corner.cjs');
const scrollAreaRoot = require('./scroll-area-root.cjs');
const scrollAreaRootProvider = require('./scroll-area-root-provider.cjs');
const scrollAreaScrollbar = require('./scroll-area-scrollbar.cjs');
const scrollAreaThumb = require('./scroll-area-thumb.cjs');
const scrollAreaViewport = require('./scroll-area-viewport.cjs');
const useScrollArea = require('./use-scroll-area.cjs');
const useScrollAreaContext = require('./use-scroll-area-context.cjs');
const scrollArea$1 = require('./scroll-area.cjs');
const scrollArea = require('@zag-js/scroll-area');



exports.ScrollAreaContext = scrollAreaContext.ScrollAreaContext;
exports.ScrollAreaContent = scrollAreaContent.ScrollAreaContent;
exports.ScrollAreaCorner = scrollAreaCorner.ScrollAreaCorner;
exports.ScrollAreaRoot = scrollAreaRoot.ScrollAreaRoot;
exports.ScrollAreaRootProvider = scrollAreaRootProvider.ScrollAreaRootProvider;
exports.ScrollAreaScrollbar = scrollAreaScrollbar.ScrollAreaScrollbar;
exports.ScrollAreaThumb = scrollAreaThumb.ScrollAreaThumb;
exports.ScrollAreaViewport = scrollAreaViewport.ScrollAreaViewport;
exports.useScrollArea = useScrollArea.useScrollArea;
exports.useScrollAreaContext = useScrollAreaContext.useScrollAreaContext;
exports.ScrollArea = scrollArea$1;
Object.defineProperty(exports, "scrollAreaAnatomy", {
  enumerable: true,
  get: () => scrollArea.anatomy
});
