// @chakra-ui/react/dist/cjs/theme/recipes.export.cjs
"use strict";
'use strict';

var badge = require('./recipes/badge.cjs');
var button = require('./recipes/button.cjs');
var checkmark = require('./recipes/checkmark.cjs');
var code = require('./recipes/code.cjs');
var colorSwatch = require('./recipes/color-swatch.cjs');
var container = require('./recipes/container.cjs');
var heading = require('./recipes/heading.cjs');
var icon = require('./recipes/icon.cjs');
var input = require('./recipes/input.cjs');
var inputAddon = require('./recipes/input-addon.cjs');
var kbd = require('./recipes/kbd.cjs');
var link = require('./recipes/link.cjs');
var mark = require('./recipes/mark.cjs');
var radiomark = require('./recipes/radiomark.cjs');
var separator = require('./recipes/separator.cjs');
var skeleton = require('./recipes/skeleton.cjs');
var skipNavLink = require('./recipes/skip-nav-link.cjs');
var spinner = require('./recipes/spinner.cjs');
var textarea = require('./recipes/textarea.cjs');



exports.badgeRecipe = badge.badgeRecipe;
exports.buttonRecipe = button.buttonRecipe;
exports.checkmarkRecipe = checkmark.checkmarkRecipe;
exports.codeRecipe = code.codeRecipe;
exports.colorSwatchRecipe = colorSwatch.colorSwatchRecipe;
exports.containerRecipe = container.containerRecipe;
exports.headingRecipe = heading.headingRecipe;
exports.iconRecipe = icon.iconRecipe;
exports.inputRecipe = input.inputRecipe;
exports.inputAddonRecipe = inputAddon.inputAddonRecipe;
exports.kbdRecipe = kbd.kbdRecipe;
exports.linkRecipe = link.linkRecipe;
exports.markRecipe = mark.markRecipe;
exports.radiomarkRecipe = radiomark.radiomarkRecipe;
exports.separatorRecipe = separator.separatorRecipe;
exports.skeletonRecipe = skeleton.skeletonRecipe;
exports.skipNavLinkRecipe = skipNavLink.skipNavLinkRecipe;
exports.spinnerRecipe = spinner.spinnerRecipe;
exports.textareaRecipe = textarea.textareaRecipe;
