// @chakra-ui/react/dist/cjs/theme/slot-recipes.export.cjs
"use strict";
'use strict';

var accordion = require('./recipes/accordion.cjs');
var actionBar = require('./recipes/action-bar.cjs');
var alert = require('./recipes/alert.cjs');
var avatar = require('./recipes/avatar.cjs');
var blockquote = require('./recipes/blockquote.cjs');
var breadcrumb = require('./recipes/breadcrumb.cjs');
var card = require('./recipes/card.cjs');
var checkbox = require('./recipes/checkbox.cjs');
var checkboxCard = require('./recipes/checkbox-card.cjs');
var codeBlock = require('./recipes/code-block.cjs');
var collapsible = require('./recipes/collapsible.cjs');
var colorPicker = require('./recipes/color-picker.cjs');
var combobox = require('./recipes/combobox.cjs');
var dataList = require('./recipes/data-list.cjs');
var dialog = require('./recipes/dialog.cjs');
var drawer = require('./recipes/drawer.cjs');
var editable = require('./recipes/editable.cjs');
var emptyState = require('./recipes/empty-state.cjs');
var field = require('./recipes/field.cjs');
var fieldset = require('./recipes/fieldset.cjs');
var fileUpload = require('./recipes/file-upload.cjs');
var hoverCard = require('./recipes/hover-card.cjs');
var list = require('./recipes/list.cjs');
var menu = require('./recipes/menu.cjs');
var nativeSelect = require('./recipes/native-select.cjs');
var numberInput = require('./recipes/number-input.cjs');
var pinInput = require('./recipes/pin-input.cjs');
var popover = require('./recipes/popover.cjs');
var progress = require('./recipes/progress.cjs');
var progressCircle = require('./recipes/progress-circle.cjs');
var qrCode = require('./recipes/qr-code.cjs');
var radioCard = require('./recipes/radio-card.cjs');
var radioGroup = require('./recipes/radio-group.cjs');
var ratingGroup = require('./recipes/rating-group.cjs');
var scrollArea = require('./recipes/scroll-area.cjs');
var segmentGroup = require('./recipes/segment-group.cjs');
var select = require('./recipes/select.cjs');
var slider = require('./recipes/slider.cjs');
var stat = require('./recipes/stat.cjs');
var status = require('./recipes/status.cjs');
var steps = require('./recipes/steps.cjs');
var _switch = require('./recipes/switch.cjs');
var table = require('./recipes/table.cjs');
var tabs = require('./recipes/tabs.cjs');
var tag = require('./recipes/tag.cjs');
var timeline = require('./recipes/timeline.cjs');
var toast = require('./recipes/toast.cjs');
var tooltip = require('./recipes/tooltip.cjs');
var treeView = require('./recipes/tree-view.cjs');



exports.accordionSlotRecipe = accordion.accordionSlotRecipe;
exports.actionBarSlotRecipe = actionBar.actionBarSlotRecipe;
exports.alertSlotRecipe = alert.alertSlotRecipe;
exports.avatarSlotRecipe = avatar.avatarSlotRecipe;
exports.blockquoteSlotRecipe = blockquote.blockquoteSlotRecipe;
exports.breadcrumbSlotRecipe = breadcrumb.breadcrumbSlotRecipe;
exports.cardSlotRecipe = card.cardSlotRecipe;
exports.checkboxSlotRecipe = checkbox.checkboxSlotRecipe;
exports.checkboxCardSlotRecipe = checkboxCard.checkboxCardSlotRecipe;
exports.codeBlockSlotRecipe = codeBlock.codeBlockSlotRecipe;
exports.collapsibleSlotRecipe = collapsible.collapsibleSlotRecipe;
exports.colorPickerSlotRecipe = colorPicker.colorPickerSlotRecipe;
exports.comboboxSlotRecipe = combobox.comboboxSlotRecipe;
exports.dataListSlotRecipe = dataList.dataListSlotRecipe;
exports.dialogSlotRecipe = dialog.dialogSlotRecipe;
exports.drawerSlotRecipe = drawer.drawerSlotRecipe;
exports.editableSlotRecipe = editable.editableSlotRecipe;
exports.emptyStateSlotRecipe = emptyState.emptyStateSlotRecipe;
exports.fieldSlotRecipe = field.fieldSlotRecipe;
exports.fieldsetSlotRecipe = fieldset.fieldsetSlotRecipe;
exports.fileUploadSlotRecipe = fileUpload.fileUploadSlotRecipe;
exports.hoverCardSlotRecipe = hoverCard.hoverCardSlotRecipe;
exports.listSlotRecipe = list.listSlotRecipe;
exports.menuSlotRecipe = menu.menuSlotRecipe;
exports.nativeSelectSlotRecipe = nativeSelect.nativeSelectSlotRecipe;
exports.numberInputSlotRecipe = numberInput.numberInputSlotRecipe;
exports.pinInputSlotRecipe = pinInput.pinInputSlotRecipe;
exports.popoverSlotRecipe = popover.popoverSlotRecipe;
exports.progressSlotRecipe = progress.progressSlotRecipe;
exports.progressCircleSlotRecipe = progressCircle.progressCircleSlotRecipe;
exports.qrCodeSlotRecipe = qrCode.qrCodeSlotRecipe;
exports.radioCardSlotRecipe = radioCard.radioCardSlotRecipe;
exports.radioGroupSlotRecipe = radioGroup.radioGroupSlotRecipe;
exports.ratingGroupSlotRecipe = ratingGroup.ratingGroupSlotRecipe;
exports.scrollAreaSlotRecipe = scrollArea.scrollAreaSlotRecipe;
exports.segmentGroupSlotRecipe = segmentGroup.segmentGroupSlotRecipe;
exports.selectSlotRecipe = select.selectSlotRecipe;
exports.sliderSlotRecipe = slider.sliderSlotRecipe;
exports.statSlotRecipe = stat.statSlotRecipe;
exports.statusSlotRecipe = status.statusSlotRecipe;
exports.stepsSlotRecipe = steps.stepsSlotRecipe;
exports.switchSlotRecipe = _switch.switchSlotRecipe;
exports.tableSlotRecipe = table.tableSlotRecipe;
exports.tabsSlotRecipe = tabs.tabsSlotRecipe;
exports.tagSlotRecipe = tag.tagSlotRecipe;
exports.timelineSlotRecipe = timeline.timelineSlotRecipe;
exports.toastSlotRecipe = toast.toastSlotRecipe;
exports.tooltipSlotRecipe = tooltip.tooltipSlotRecipe;
exports.treeViewSlotRecipe = treeView.treeViewSlotRecipe;
