// @chakra-ui/react/dist/cjs/components/scroll-area/index.cjs
"use strict";
'use strict';

var scrollArea = require('./scroll-area.cjs');
var scrollArea$1 = require('@ark-ui/react/scroll-area');
var namespace = require('./namespace.cjs');



exports.ScrollAreaContent = scrollArea.ScrollAreaContent;
exports.ScrollAreaContext = scrollArea.ScrollAreaContext;
exports.ScrollAreaCorner = scrollArea.ScrollAreaCorner;
exports.ScrollAreaPropsProvider = scrollArea.ScrollAreaPropsProvider;
exports.ScrollAreaRoot = scrollArea.ScrollAreaRoot;
exports.ScrollAreaRootProvider = scrollArea.ScrollAreaRootProvider;
exports.ScrollAreaScrollbar = scrollArea.ScrollAreaScrollbar;
exports.ScrollAreaThumb = scrollArea.ScrollAreaThumb;
exports.ScrollAreaViewport = scrollArea.ScrollAreaViewport;
exports.useScrollAreaStyles = scrollArea.useScrollAreaStyles;
Object.defineProperty(exports, "useScrollArea", {
  enumerable: true,
  get: function () { return scrollArea$1.useScrollArea; }
});
Object.defineProperty(exports, "useScrollAreaContext", {
  enumerable: true,
  get: function () { return scrollArea$1.useScrollAreaContext; }
});
exports.ScrollArea = namespace;
