// @chakra-ui/react/dist/cjs/components/code-block/index.cjs
"use strict";
'use strict';

var codeBlockAdapterProvider = require('./code-block-adapter-provider.cjs');
var codeBlock = require('./code-block.cjs');
var adapters = require('./adapters.cjs');
var namespace = require('./namespace.cjs');



exports.CodeBlockAdapterProvider = codeBlockAdapterProvider.CodeBlockAdapterProvider;
exports.CodeBlockCode = codeBlock.CodeBlockCode;
exports.CodeBlockCodeText = codeBlock.CodeBlockCodeText;
exports.CodeBlockCollapseIndicator = codeBlock.CodeBlockCollapseIndicator;
exports.CodeBlockCollapseText = codeBlock.CodeBlockCollapseText;
exports.CodeBlockCollapseTrigger = codeBlock.CodeBlockCollapseTrigger;
exports.CodeBlockContent = codeBlock.CodeBlockContent;
exports.CodeBlockContext = codeBlock.CodeBlockContext;
exports.CodeBlockControl = codeBlock.CodeBlockControl;
exports.CodeBlockCopyIndicator = codeBlock.CodeBlockCopyIndicator;
exports.CodeBlockCopyTrigger = codeBlock.CodeBlockCopyTrigger;
exports.CodeBlockFooter = codeBlock.CodeBlockFooter;
exports.CodeBlockHeader = codeBlock.CodeBlockHeader;
exports.CodeBlockOverlay = codeBlock.CodeBlockOverlay;
exports.CodeBlockRoot = codeBlock.CodeBlockRoot;
exports.CodeBlockTitle = codeBlock.CodeBlockTitle;
exports.createHighlightJsAdapter = adapters.createHighlightJsAdapter;
exports.createShikiAdapter = adapters.createShikiAdapter;
exports.plainTextAdapter = adapters.plainTextAdapter;
exports.CodeBlock = namespace;
